File: app/src/main/java/piuk/blockchain/android/ui/transactions/TransactionDetailPresenter.java
package piuk.blockchain.android.ui.transactions;

import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;
import info.blockchain.balance.CryptoCurrency;
import info.blockchain.balance.FiatValue;
import info.blockchain.wallet.contacts.data.FacilitatedTransaction;
import info.blockchain.wallet.multiaddress.MultiAddressFactory;
import info.blockchain.wallet.multiaddress.TransactionSummary.Direction;
import info.blockchain.wallet.util.FormatsUtil;
import io.reactivex.Completable;
import io.reactivex.Observable;
import org.apache.commons.lang3.tuple.Pair;
import piuk.blockchain.android.R;
import piuk.blockchain.android.data.bitcoincash.BchDataManager;
import piuk.blockchain.androidcore.data.contacts.models.ContactTransactionDisplayModel;
import piuk.blockchain.android.data.datamanagers.TransactionListDataManager;
import piuk.blockchain.android.data.ethereum.EthDataManager;
import piuk.blockchain.android.data.rxjava.RxUtil;
import piuk.blockchain.android.util.StringUtils;
import piuk.blockchain.androidcore.data.api.EnvironmentConfig;
import piuk.blockchain.androidcore.data.contacts.ContactsDataManager;
import piuk.blockchain.androidcore.data.currency.BTCDenomination;
import piuk.blockchain.androidcore.data.currency.CurrencyFormatManager;
import piuk.blockchain.androidcore.data.currency.ETHDenomination;
import piuk.blockchain.androidcore.data.exchangerate.ExchangeRateDataManager;
import piuk.blockchain.androidcore.data.payload.PayloadDataManager;
import piuk.blockchain.androidcore.data.transactions.models.Displayable;
import piuk.blockchain.androidcore.utils.PrefsUtil;
import piuk.blockchain.androidcoreui.ui.base.BasePresenter;
import piuk.blockchain.androidcoreui.ui.customviews.ToastCustom;
import timber.log.Timber;

import javax.inject.Inject;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

import static piuk.blockchain.android.ui.balance.BalanceFragment.KEY_TRANSACTION_HASH;
import static piuk.blockchain.android.ui.balance.BalanceFragment.KEY_TRANSACTION_LIST_POSITION;

@SuppressWarnings("WeakerAccess")
public class TransactionDetailPresenter extends BasePresenter<TransactionDetailView> {

    private TransactionHelper transactionHelper;
    private PayloadDataManager payloadDataManager;
    private StringUtils stringUtils;
    private TransactionListDataManager transactionListDataManager;
    private ExchangeRateDataManager exchangeRateFactory;
    private ContactsDataManager contactsDataManager;
    private EthDataManager ethDataManager;
    private BchDataManager bchDataManager;
    private EnvironmentConfig environmentSettings;
    private CurrencyFormatManager currencyFormatManager;

    private String fiatType;

    @VisibleForTesting
    Displayable displayable;

    @Inject
    public TransactionDetailPresenter(TransactionHelper transactionHelper,
                                      PrefsUtil prefsUtil,
                                      PayloadDataManager payloadDataManager,
                                      StringUtils stringUtils,
                                      TransactionListDataManager transactionListDataManager,
                                      ExchangeRateDataManager exchangeRateFactory,
                                      ContactsDataManager contactsDataManager,
                                      EthDataManager ethDataManager,
                                      BchDataManager bchDataManager,
                                      EnvironmentConfig environmentSettings,
                                      CurrencyFormatManager currencyFormatManager) {

        this.transactionHelper = transactionHelper;
        fiatType = prefsUtil.getValue(PrefsUtil.KEY_SELECTED_FIAT, PrefsUtil.DEFAULT_CURRENCY);
        this.payloadDataManager = payloadDataManager;
        this.stringUtils = stringUtils;
        this.transactionListDataManager = transactionListDataManager;
        this.exchangeRateFactory = exchangeRateFactory;
        this.contactsDataManager = contactsDataManager;
        this.ethDataManager = ethDataManager;
        this.bchDataManager = bchDataManager;
        this.environmentSettings = environmentSettings;
        this.currencyFormatManager = currencyFormatManager;
    }

    @Override
    public void onViewReady() {
        Intent pageIntent = getView().getPageIntent();
        if (pageIntent != null && pageIntent.hasExtra(KEY_TRANSACTION_LIST_POSITION)) {
            int transactionPosition = pageIntent.getIntExtra(KEY_TRANSACTION_LIST_POSITION, -1);
            if (transactionPosition != -1) {
                displayable = transactionListDataManager.getTransactionList().get(transactionPosition);
                updateUiFromTransaction(displayable);
            } else {
                getView().pageFinish();
            }
        } else if (pageIntent != null && pageIntent.hasExtra(KEY_TRANSACTION_HASH)) {
            getCompositeDisposable().add(
                    transactionListDataManager.getTxFromHash(pageIntent.getStringExtra(KEY_TRANSACTION_HASH))
                            .doOnSuccess(displayable -> this.displayable = displayable)
                            .subscribe(
                                    this::updateUiFromTransaction,
                                    throwable -> getView().pageFinish()));
        } else {
            Timber.e("Transaction hash not found");
            getView().pageFinish();
        }
    }

    void updateTransactionNote(String description) {
        Completable completable;
        if (displayable.getCryptoCurrency() == CryptoCurrency.BTC) {
            completable = payloadDataManager.updateTransactionNotes(displayable.getHash(), description);
        } else if (displayable.getCryptoCurrency() == CryptoCurrency.ETHER) {
            completable = ethDataManager.updateTransactionNotes(displayable.getHash(), description);
        } else {
            throw new IllegalArgumentException("Only BTC and ETHER currently supported");
        }

        completable.compose(RxUtil.addCompletableToCompositeDisposable(this))
                .subscribe(() -> {
                    getView().showToast(R.string.remote_save_ok, ToastCustom.TYPE_OK);
                    getView().setDescription(description);
                }, throwable -> getView().showToast(R.string.unexpected_error, ToastCustom.TYPE_ERROR));
    }

    private void updateUiFromTransaction(Displayable displayable) {
        getView().setTransactionType(displayable.getDirection());
        setTransactionColor(displayable);
        setTransactionAmountInBtcOrEth(displayable.getCryptoCurrency(), displayable.getTotal());
        setConfirmationStatus(displayable.getCryptoCurrency(), displayable.getHash(), displayable.getConfirmations());
        setTransactionNote(displayable.getHash());
        setDate(displayable.getTimeStamp());
        setFee(displayable.getCryptoCurrency(), displayable.getFee());

        if (displayable.getCryptoCurrency() == CryptoCurrency.BTC) {
            handleBtcToAndFrom(displayable);
        } else if (displayable.getCryptoCurrency() == CryptoCurrency.ETHER) {
            handleEthToAndFrom(displayable);
        } else if (displayable.getCryptoCurrency() == CryptoCurrency.BCH) {
            handleBchToAndFrom(displayable);
        } else {
            throw new IllegalArgumentException(displayable.getCryptoCurrency() + " is not currently supported");
        }

        getCompositeDisposable().add(
                getTransactionValueString(fiatType, displayable)
                        .subscribe(
                                value -> getView().setTransactionValueFiat(value),
                                throwable -> getView().showToast(R.string.unexpected_error, ToastCustom.TYPE_ERROR)));

        getView().onDataLoaded();
        getView().setIsDoubleSpend(displayable.getDoubleSpend());
    }

    @SuppressWarnings("ConstantConditions")
    private void handleEthToAndFrom(Displayable displayable) {
        String fromAddress = displayable.getInputsMap().keySet().iterator().next();
        String toAddress = displayable.getOutputsMap().keySet().iterator().next();

        String ethAddress = ethDataManager.getEthResponseModel().getAddressResponse().getAccount();
        if (fromAddress.equals(ethAddress)) {
            fromAddress = stringUtils.getString(R.string.eth_default_account_label);
        }
        if (toAddress.equals(ethAddress)) {
            toAddress = stringUtils.getString(R.string.eth_default_account_label);
        }

        getView().setFromAddress(Collections.singletonList(new TransactionDetailModel(
                fromAddress, "", "")));
        getView().setToAddresses(Collections.singletonList(new TransactionDetailModel(
                toAddress, "", "")));
    }

    private void handleBtcToAndFrom(Displayable displayable) {
        Pair<HashMap<String, BigInteger>, HashMap<String, BigInteger>> pair =
                transactionHelper.filterNonChangeAddresses(displayable);

        // From Addresses
        ArrayList<TransactionDetailModel> fromList = getFromList(displayable.getCryptoCurrency(), pair.getLeft());
        getView().setFromAddress(fromList);

        // From Contacts
        ContactTransactionDisplayModel displayModel = null;
        if (contactsDataManager.getTransactionDisplayMap().containsKey(displayable.getHash())) {
            displayModel = contactsDataManager.getTransactionDisplayMap().get(displayable.getHash());

            // Check if should be "Paid" state via contacts
            if (displayModel != null) {
                if (displayModel.getState().equals(FacilitatedTransaction.STATE_PAYMENT_BROADCASTED)
                        && displayModel.getRole().equals(FacilitatedTransaction.ROLE_PR_RECEIVER)) {
                    getView().showTransactionAsPaid();
                }
            }
        }

        // To Addresses
        ArrayList<TransactionDetailModel> recipients = getToList(displayable.getCryptoCurrency(), displayModel, pair.getRight());
        getView().setToAddresses(recipients);

        if (displayModel != null) {
            getView().setTransactionNote(displayModel.getNote());
        }
    }

    private void handleBchToAndFrom(Displayable displayable) {
        Pair<HashMap<String, BigInteger>, HashMap<String, BigInteger>> pair =
                transactionHelper.filterNonChangeAddressesBch(displayable);

        // From Addresses
        ArrayList<TransactionDetailModel> fromList = getFromList(displayable.getCryptoCurrency(), pair.getLeft());
        getView().setFromAddress(fromList);

        // From Contacts
        ContactTransactionDisplayModel displayModel = null;
        if (contactsDataManager.getTransactionDisplayMap().containsKey(displayable.getHash())) {
            displayModel = contactsDataManager.getTransactionDisplayMap().get(displayable.getHash());

            // Check if should be "Paid" state via contacts
            if (displayModel != null) {
                if (displayModel.getState().equals(FacilitatedTransaction.STATE_PAYMENT_BROADCASTED)
                        && displayModel.getRole().equals(FacilitatedTransaction.ROLE_PR_RECEIVER)) {
                    getView().showTransactionAsPaid();
                }
            }
        }

        // To Addresses
        ArrayList<TransactionDetailModel> recipients = getToList(displayable.getCryptoCurrency(), displayModel, pair.getRight());
        getView().setToAddresses(recipients);

        if (displayModel != null) {
            getView().setTransactionNote(displayModel.getNote());
        }
    }

    private ArrayList<TransactionDetailModel> getFromList(CryptoCurrency currency, HashMap<String, BigInteger> inputMap) {
        ArrayList<TransactionDetailModel> inputs = new ArrayList<>();

        String unit = "";
        for (Map.Entry<String, BigInteger> item : inputMap.entrySet()) {

            long value = (item.getValue() != null) ? item.getValue().longValue() : 0;

            String label;
            if (currency == CryptoCurrency.BTC) {
                label = payloadDataManager.addressToLabel(item.getKey());
                unit = getDisplayUnitsBtc();
            } else {
                label = bchDataManager.getLabelFromBchAddress(item.getKey());
                unit = getDisplayUnitsBch();
                if (label == null) label = FormatsUtil.toShortCashAddress(
                        environmentSettings.getBitcoinCashNetworkParameters(),
                        item.getKey());
            }

            TransactionDetailModel transactionDetailModel = new TransactionDetailModel(
                    label,
                    currencyFormatManager.getFormattedSelectedCoinValue(BigInteger.valueOf(value)),
                    unit);

            if (transactionDetailModel.getAddress().equals(MultiAddressFactory.ADDRESS_DECODE_ERROR)) {
                transactionDetailModel.setAddress(stringUtils.getString(R.string.tx_decode_error));
                transactionDetailModel.setAddressDecodeError(true);
            }

            inputs.add(transactionDetailModel);
        }

        //No inputs = coinbase
        if (inputs.isEmpty()) {
            TransactionDetailModel coinbase = new TransactionDetailModel(
                    stringUtils.getString(R.string.transaction_detail_coinbase),
                    "",
                    unit);

            inputs.add(coinbase);
        }

        return inputs;
    }

    private ArrayList<TransactionDetailModel> getToList(CryptoCurrency currency,
                                                        ContactTransactionDisplayModel displayModel,
                                                        HashMap<String, BigInteger> outputMap) {
        ArrayList<TransactionDetailModel> recipients = new ArrayList<>();

        for (Map.Entry<String, BigInteger> item : outputMap.entrySet()) {

            long value = (item.getValue() != null) ? item.getValue().longValue() : 0;
            String unit;
            String label;

            if (currency == CryptoCurrency.BTC) {
                label = payloadDataManager.addressToLabel(item.getKey());
                unit = getDisplayUnitsBtc();
            } else {
                label = bchDataManager.getLabelFromBchAddress(item.getKey());
                if (label == null) label = FormatsUtil.toShortCashAddress(
                        environmentSettings.getBitcoinCashNetworkParameters(),
                        item.getKey());
                unit = getDisplayUnitsBch();
            }

            TransactionDetailModel transactionDetailModel = new TransactionDetailModel(
                    label,
                    currencyFormatManager.getFormattedSelectedCoinValue(BigInteger.valueOf(value)),
                    unit);

            if (displayModel != null && displayable.getDirection().equals(Direction.SENT)) {
                transactionDetailModel.setAddress(displayModel.getContactName());
            }

            if (transactionDetailModel.getAddress().equals(MultiAddressFactory.ADDRESS_DECODE_ERROR)) {
                transactionDetailModel.setAddress(stringUtils.getString(R.string.tx_decode_error));
                transactionDetailModel.setAddressDecodeError(true);
            }

            recipients.add(transactionDetailModel);
        }

        return recipients;
    }

    private void setFee(CryptoCurrency currency, BigInteger fee) {
        if (currency == CryptoCurrency.BTC) {
            String formattedFee = (
                    currencyFormatManager.getFormattedBtcValueWithUnit(
                            new BigDecimal(fee),
                            BTCDenomination.SATOSHI));
            getView().setFee(formattedFee);
        } else if (currency == CryptoCurrency.ETHER) {
            String formattedFee = (
                    currencyFormatManager.getFormattedEthShortValueWithUnit(
                            new BigDecimal(fee),
                            ETHDenomination.WEI));
            getView().setFee(formattedFee);
        } else if (currency == CryptoCurrency.BCH) {
            String formattedFee = (
                    currencyFormatManager.getFormattedBchValueWithUnit(
                            new BigDecimal(fee),
                            BTCDenomination.SATOSHI));
            getView().setFee(formattedFee);
        } else {
            throw new IllegalArgumentException(currency + " is not currently supported");
        }
    }

    private void setTransactionAmountInBtcOrEth(CryptoCurrency currency, BigInteger total) {
        if (currency == CryptoCurrency.ETHER) {
            String amountEth = (
                    currencyFormatManager.getFormattedEthShortValueWithUnit(
                            new BigDecimal(total.abs()), ETHDenomination.WEI));

            getView().setTransactionValueBtc(amountEth);
        } else if (currency == CryptoCurrency.BTC) {
            String amountBtc = (
                    currencyFormatManager.getFormattedBtcValueWithUnit(
                            new BigDecimal(total.abs()), BTCDenomination.SATOSHI));

            getView().setTransactionValueBtc(amountBtc);
        } else {
            String amountBch = (
                    currencyFormatManager.getFormattedBchValueWithUnit(
                            new BigDecimal(total.abs()), BTCDenomination.SATOSHI));

            getView().setTransactionValueBtc(amountBch);
        }
    }

    private void setTransactionNote(String txHash) {
        String notes;
        if (displayable.getCryptoCurrency() == CryptoCurrency.BTC) {
            notes = payloadDataManager.getTransactionNotes(txHash);
        } else if (displayable.getCryptoCurrency() == CryptoCurrency.ETHER) {
            notes = ethDataManager.getTransactionNotes(displayable.getHash());
        } else {
            //Only BTC and ETHER currently supported
            notes = "";
            getView().hideDescriptionField();
        }
        getView().setDescription(notes);
    }

    public String getTransactionNote() {
        if (displayable.getCryptoCurrency() == CryptoCurrency.BTC) {
            return payloadDataManager.getTransactionNotes(displayable.getHash());
        } else if (displayable.getCryptoCurrency() == CryptoCurrency.ETHER) {
            return ethDataManager.getTransactionNotes(displayable.getHash());
        } else {
            // Currently no available notes for bch
            return "";
        }
    }

    public String getTransactionHash() {
        return displayable.getHash();
    }

    public CryptoCurrency getTransactionType() {
        return displayable.getCryptoCurrency();
    }

    @VisibleForTesting
    void setConfirmationStatus(CryptoCurrency cryptoCurrency, String txHash, long confirmations) {
        if (confirmations >= cryptoCurrency.getRequiredConfirmations()) {
            getView().setStatus(cryptoCurrency, stringUtils.getString(R.string.transaction_detail_confirmed), txHash);
        } else {
            String pending = stringUtils.getString(R.string.transaction_detail_pending);
            pending = String.format(Locale.getDefault(), pending, confirmations, cryptoCurrency.getRequiredConfirmations());
            getView().setStatus(cryptoCurrency, pending, txHash);
        }
    }

    private void setDate(long time) {
        long epochTime = time * 1000;

        Date date = new Date(epochTime);
        DateFormat dateFormat = SimpleDateFormat.getDateInstance(DateFormat.LONG);
        DateFormat timeFormat = new SimpleDateFormat("hh:mm a", Locale.getDefault());
        String dateText = dateFormat.format(date);
        String timeText = timeFormat.format(date);

        getView().setDate(dateText + " @ " + timeText);
    }

    @VisibleForTesting
    void setTransactionColor(Displayable transaction) {
        if (transaction.getDirection() == Direction.TRANSFERRED) {
            getView().setTransactionColour(transaction.getConfirmations() < transaction.getCryptoCurrency().getRequiredConfirmations()
                    ? R.color.product_gray_transferred_50 : R.color.product_gray_transferred);
        } else if (transaction.getDirection() == Direction.SENT) {
            getView().setTransactionColour(transaction.getConfirmations() < transaction.getCryptoCurrency().getRequiredConfirmations()
                    ? R.color.product_red_sent_50 : R.color.product_red_sent);
        } else {
            getView().setTransactionColour(transaction.getConfirmations() < transaction.getCryptoCurrency().getRequiredConfirmations()
                    ? R.color.product_green_received_50 : R.color.product_green_received);
        }
    }

    @VisibleForTesting
    Observable<String> getTransactionValueString(String currency, Displayable transaction) {
        if (transaction.getCryptoCurrency() == CryptoCurrency.BTC) {
            return exchangeRateFactory.getBtcHistoricPrice(
                    transaction.getTotal().longValue(),
                    currency,
                    transaction.getTimeStamp())
                    .map(aDouble -> getTransactionString(transaction, aDouble));
        } else if (transaction.getCryptoCurrency() == CryptoCurrency.BCH) {
            return exchangeRateFactory.getBchHistoricPrice(
                    transaction.getTotal().longValue(),
                    currency,
                    transaction.getTimeStamp())
                    .map(aDouble -> getTransactionString(transaction, aDouble));
        } else {
            return exchangeRateFactory.getEthHistoricPrice(
                    transaction.getTotal(),
                    currency,
                    transaction.getTimeStamp())
                    .map(aDouble -> getTransactionString(transaction, aDouble));
        }
    }

    @NonNull
    private String getTransactionString(Displayable transaction, BigDecimal aDouble) {
        int stringId = -1;
        switch (transaction.getDirection()) {
            case TRANSFERRED:
                stringId = R.string.transaction_detail_value_at_time_transferred;
                break;
            case SENT:
                stringId = R.string.transaction_detail_value_at_time_sent;
                break;
            case RECEIVED:
                stringId = R.string.transaction_detail_value_at_time_received;
                break;
        }
        return stringUtils.getString(stringId) +
                FiatValue.fromMajor(fiatType, aDouble).toStringWithSymbol(Locale.getDefault());
    }

    private String getDisplayUnitsBtc() {
        return CryptoCurrency.BTC.name();
    }

    private String getDisplayUnitsBch() {
        return CryptoCurrency.BCH.name();
    }
}

File: app/src/main/java/piuk/blockchain/android/ui/transactions/TransactionDetailPresenter.kt
package piuk.blockchain.android.ui.transactions

import android.support.annotation.VisibleForTesting
import com.blockchain.data.TransactionHash
import com.blockchain.sunriver.XlmDataManager
import info.blockchain.balance.CryptoCurrency
import info.blockchain.balance.CryptoValue
import info.blockchain.balance.FiatValue
import info.blockchain.balance.format
import info.blockchain.balance.formatWithUnit
import info.blockchain.wallet.multiaddress.MultiAddressFactory
import info.blockchain.wallet.multiaddress.TransactionSummary
import info.blockchain.wallet.util.FormatsUtil
import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.rxkotlin.plusAssign
import io.reactivex.rxkotlin.subscribeBy
import piuk.blockchain.android.R
import piuk.blockchain.android.data.datamanagers.TransactionListDataManager
import piuk.blockchain.android.ui.balance.BalanceFragment.Companion.KEY_TRANSACTION_HASH
import piuk.blockchain.android.ui.balance.BalanceFragment.Companion.KEY_TRANSACTION_LIST_POSITION
import piuk.blockchain.android.ui.balance.adapter.formatting
import piuk.blockchain.android.util.StringUtils
import piuk.blockchain.androidcore.data.api.EnvironmentConfig
import piuk.blockchain.androidcore.data.bitcoincash.BchDataManager
import piuk.blockchain.androidcore.data.ethereum.EthDataManager
import piuk.blockchain.androidcore.data.exchangerate.ExchangeRateDataManager
import piuk.blockchain.androidcore.data.payload.PayloadDataManager
import piuk.blockchain.androidcore.data.transactions.models.Displayable
import piuk.blockchain.androidcore.utils.PrefsUtil
import piuk.blockchain.androidcoreui.ui.base.BasePresenter
import piuk.blockchain.androidcoreui.ui.customviews.ToastCustom
import timber.log.Timber
import java.math.BigInteger
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.HashMap
import java.util.Locale
import javax.inject.Inject

class TransactionDetailPresenter @Inject constructor(
    private val transactionHelper: TransactionHelper,
    prefsUtil: PrefsUtil,
    private val payloadDataManager: PayloadDataManager,
    private val stringUtils: StringUtils,
    private val transactionListDataManager: TransactionListDataManager,
    private val exchangeRateDataManager: ExchangeRateDataManager,
    private val ethDataManager: EthDataManager,
    private val bchDataManager: BchDataManager,
    private val environmentSettings: EnvironmentConfig,
    private val xlmDataManager: XlmDataManager
) : BasePresenter<TransactionDetailView>() {

    private val fiatType: String = prefsUtil.getValue(PrefsUtil.KEY_SELECTED_FIAT, PrefsUtil.DEFAULT_CURRENCY)

    @VisibleForTesting
    lateinit var displayable: Displayable

    // Currently no available notes for bch
    // Only BTC and ETHER currently supported
    var transactionNote: String?
        get() = when {
            displayable.cryptoCurrency == CryptoCurrency.BTC -> payloadDataManager.getTransactionNotes(displayable.hash)
            displayable.cryptoCurrency == CryptoCurrency.ETHER -> ethDataManager.getTransactionNotes(displayable.hash)
            else -> ""
        }
        private set(txHash) {
            val notes: String? = when (displayable.cryptoCurrency) {
                CryptoCurrency.BTC -> payloadDataManager.getTransactionNotes(txHash!!)
                CryptoCurrency.ETHER -> ethDataManager.getTransactionNotes(displayable.hash)
                else -> {
                    view.hideDescriptionField()
                    ""
                }
            }
            view.setDescription(notes)
        }

    val transactionHash: TransactionHash
        get() = TransactionHash(displayable.cryptoCurrency, displayable.hash)

    val transactionType: CryptoCurrency
        get() = displayable.cryptoCurrency

    override fun onViewReady() {
        val pageIntent = view.getPageIntent()
        if (pageIntent != null && pageIntent.hasExtra(KEY_TRANSACTION_LIST_POSITION)) {
            val transactionPosition = pageIntent.getIntExtra(KEY_TRANSACTION_LIST_POSITION, -1)
            if (transactionPosition != -1) {
                displayable = transactionListDataManager.getTransactionList()[transactionPosition]
                updateUiFromTransaction(displayable)
            } else {
                view.pageFinish()
            }
        } else if (pageIntent != null && pageIntent.hasExtra(KEY_TRANSACTION_HASH)) {
            compositeDisposable +=
                transactionListDataManager.getTxFromHash(pageIntent.getStringExtra(KEY_TRANSACTION_HASH))
                    .doOnSuccess { displayable = it }
                    .subscribe(
                        { this.updateUiFromTransaction(it) },
                        { view.pageFinish() })
        } else {
            Timber.e("Transaction hash not found")
            view.pageFinish()
        }
    }

    fun updateTransactionNote(description: String) {
        val completable: Completable = when (displayable.cryptoCurrency) {
            CryptoCurrency.BTC -> payloadDataManager.updateTransactionNotes(
                displayable.hash,
                description
            )
            CryptoCurrency.ETHER -> ethDataManager.updateTransactionNotes(
                displayable.hash,
                description
            )
            else -> throw IllegalArgumentException("Only BTC and ETHER currently supported")
        }

        compositeDisposable +=
            completable.subscribe(
                {
                    view.showToast(R.string.remote_save_ok, ToastCustom.TYPE_OK)
                    view.setDescription(description)
                },
                { view.showToast(R.string.unexpected_error, ToastCustom.TYPE_ERROR) }
            )
    }

    private fun updateUiFromTransaction(displayable: Displayable) {
        with(displayable) {
            view.setTransactionType(direction)
            setTransactionColor(this)
            setTransactionValue(cryptoCurrency, total)
            setConfirmationStatus(cryptoCurrency, hash, confirmations.toLong())
            transactionNote = hash
            setDate(timeStamp)
            setTransactionFee(cryptoCurrency, fee)
            if (cryptoCurrency == CryptoCurrency.XLM) fetchXlmFee(hash)

            when (cryptoCurrency) {
                CryptoCurrency.BTC -> handleBtcToAndFrom(this)
                CryptoCurrency.ETHER -> handleEthToAndFrom(this)
                CryptoCurrency.BCH -> handleBchToAndFrom(this)
                CryptoCurrency.XLM -> handleXlmToAndFrom(this)
                else -> throw IllegalArgumentException(cryptoCurrency.toString() + " is not currently supported")
            }

            compositeDisposable +=
                getTransactionValueString(fiatType, this)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                        { value -> view.setTransactionValueFiat(value) },
                        { view.showToast(R.string.unexpected_error, ToastCustom.TYPE_ERROR) })

            view.onDataLoaded()
            view.setIsDoubleSpend(doubleSpend)
        }
    }

    private fun handleXlmToAndFrom(displayable: Displayable) {
        compositeDisposable +=
            xlmDataManager.defaultAccount()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                    onSuccess = {
                        var fromAddress = displayable.inputsMap.keys.first()
                        var toAddress = displayable.outputsMap.keys.first()
                        if (fromAddress == it.accountId) {
                            fromAddress = stringUtils.getString(R.string.xlm_default_account_label)
                        }
                        if (toAddress == it.accountId) {
                            toAddress = stringUtils.getString(R.string.xlm_default_account_label)
                        }

                        view.setFromAddress(listOf(TransactionDetailModel(fromAddress, "", "")))
                        view.setToAddresses(listOf(TransactionDetailModel(toAddress, "", "")))
                    }
                )
    }

    private fun handleEthToAndFrom(displayable: Displayable) {
        var fromAddress = displayable.inputsMap.keys.first()
        var toAddress = displayable.outputsMap.keys.first()

        val ethAddress = ethDataManager.getEthResponseModel()!!.getAddressResponse()!!.account
        if (fromAddress == ethAddress) {
            fromAddress = stringUtils.getString(R.string.eth_default_account_label)
        }
        if (toAddress == ethAddress) {
            toAddress = stringUtils.getString(R.string.eth_default_account_label)
        }

        view.setFromAddress(listOf(TransactionDetailModel(fromAddress, "", "")))
        view.setToAddresses(listOf(TransactionDetailModel(toAddress, "", "")))
    }

    private fun handleBtcToAndFrom(displayable: Displayable) {
        val (inputs, outputs) = transactionHelper.filterNonChangeAddresses(displayable)
        setToAndFrom(displayable, inputs, outputs)
    }

    private fun handleBchToAndFrom(displayable: Displayable) {
        val (inputs, outputs) = transactionHelper.filterNonChangeAddressesBch(displayable)
        setToAndFrom(displayable, inputs, outputs)
    }

    private fun setToAndFrom(
        displayable: Displayable,
        inputs: HashMap<String, BigInteger?>,
        outputs: HashMap<String, BigInteger?>
    ) {
        // From Addresses
        val fromList = getFromList(displayable.cryptoCurrency, inputs)
        view.setFromAddress(fromList)
        // To Addresses
        val recipients = getToList(displayable.cryptoCurrency, outputs)
        view.setToAddresses(recipients)
    }

    private fun getFromList(
        currency: CryptoCurrency,
        inputMap: HashMap<String, BigInteger?>
    ): List<TransactionDetailModel> {
        val inputs = handleTransactionMap(inputMap, currency)
        // No inputs = coinbase transaction
        if (inputs.isEmpty()) {
            val coinbase = TransactionDetailModel(
                stringUtils.getString(R.string.transaction_detail_coinbase),
                "",
                currency.symbol
            )

            inputs.add(coinbase)
        }

        return inputs.toList()
    }

    private fun getToList(
        currency: CryptoCurrency,
        outputMap: HashMap<String, BigInteger?>
    ): List<TransactionDetailModel> = handleTransactionMap(outputMap, currency)

    private fun handleTransactionMap(
        inputMap: HashMap<String, BigInteger?>,
        currency: CryptoCurrency
    ): MutableList<TransactionDetailModel> {
        val inputs = mutableListOf<TransactionDetailModel>()
        for ((key, value) in inputMap) {
            var label: String?
            if (currency == CryptoCurrency.BTC) {
                label = payloadDataManager.addressToLabel(key)
            } else {
                label = bchDataManager.getLabelFromBchAddress(key)
                if (label == null)
                    label = FormatsUtil.toShortCashAddress(
                        environmentSettings.bitcoinCashNetworkParameters,
                        key
                    )
            }

            val transactionDetailModel = buildTransactionDetailModel(label, currency, value, currency.symbol)
            inputs.add(transactionDetailModel)
        }
        return inputs
    }

    private fun buildTransactionDetailModel(
        label: String?,
        currency: CryptoCurrency,
        value: BigInteger?,
        unit: String
    ): TransactionDetailModel = TransactionDetailModel(
        label,
        if (currency == CryptoCurrency.BTC) {
            CryptoValue.bitcoinFromSatoshis(value ?: BigInteger.ZERO).format()
        } else {
            CryptoValue.bitcoinCashFromSatoshis(value ?: BigInteger.ZERO).format()
        },
        unit
    ).apply {
        if (address == MultiAddressFactory.ADDRESS_DECODE_ERROR) {
            address = stringUtils.getString(R.string.tx_decode_error)
            setAddressDecodeError(true)
        }
    }

    private fun fetchXlmFee(hash: String) {
        compositeDisposable +=
            xlmDataManager.getTransactionFee(hash)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                    onSuccess = { view.setFee(it.formatWithUnit()) }
                )
    }

    private fun setTransactionFee(currency: CryptoCurrency, fee: BigInteger) {
        when (currency) {
            CryptoCurrency.BTC -> CryptoValue.bitcoinFromSatoshis(fee)
            CryptoCurrency.ETHER -> CryptoValue.etherFromWei(fee)
            CryptoCurrency.BCH -> CryptoValue.bitcoinCashFromSatoshis(fee)
            CryptoCurrency.XLM -> CryptoValue.lumensFromStroop(fee)
            else -> throw IllegalArgumentException("$currency is not currently supported")
        }.run { view.setFee(this.formatWithUnit()) }
    }

    private fun setTransactionValue(currency: CryptoCurrency, total: BigInteger) {
        when (currency) {
            CryptoCurrency.ETHER -> CryptoValue.etherFromWei(total)
            CryptoCurrency.BTC -> CryptoValue.bitcoinFromSatoshis(total)
            CryptoCurrency.BCH -> CryptoValue.bitcoinCashFromSatoshis(total)
            CryptoCurrency.XLM -> CryptoValue.lumensFromStroop(total)
            else -> throw IllegalArgumentException("$currency is not currently supported")
        }.run { view.setTransactionValue(this.formatWithUnit()) }
    }

    @VisibleForTesting
    internal fun setConfirmationStatus(cryptoCurrency: CryptoCurrency, txHash: String, confirmations: Long) {
        if (confirmations >= cryptoCurrency.requiredConfirmations) {
            view.setStatus(cryptoCurrency, stringUtils.getString(R.string.transaction_detail_confirmed), txHash)
        } else {
            var pending = stringUtils.getString(R.string.transaction_detail_pending)
            pending = String.format(Locale.getDefault(), pending, confirmations, cryptoCurrency.requiredConfirmations)
            view.setStatus(cryptoCurrency, pending, txHash)
        }
    }

    private fun setDate(time: Long) {
        val epochTime = time * 1000

        val date = Date(epochTime)
        val dateFormat = SimpleDateFormat.getDateInstance(DateFormat.LONG)
        val timeFormat = SimpleDateFormat("hh:mm a", Locale.getDefault())
        val dateText = dateFormat.format(date)
        val timeText = timeFormat.format(date)

        view.setDate("$dateText @ $timeText")
    }

    @VisibleForTesting
    internal fun setTransactionColor(transaction: Displayable) {
        view.setTransactionColour(transaction.formatting().directionColour)
    }

    @VisibleForTesting
    internal fun getTransactionValueString(currency: String, transaction: Displayable): Single<String> =
        exchangeRateDataManager.getHistoricPrice(
            CryptoValue(transaction.cryptoCurrency, transaction.total),
            currency,
            transaction.timeStamp
        ).map { getTransactionString(transaction, it) }

    private fun getTransactionString(transaction: Displayable, value: FiatValue): String {
        val stringId = when (transaction.direction) {
            TransactionSummary.Direction.TRANSFERRED -> R.string.transaction_detail_value_at_time_transferred
            TransactionSummary.Direction.SENT -> R.string.transaction_detail_value_at_time_sent
            TransactionSummary.Direction.RECEIVED -> R.string.transaction_detail_value_at_time_received
        }
        return stringUtils.getString(stringId) + value.toStringWithSymbol()
    }
}
